internal/git/repository.go

```diff
@@ -94,38 +98,73 @@ func NewRepository(fs afero.Fs, git *CommandExecutor) (*Repository, error) {
 		}
 	}
 
-	gitPath, err := git.Cmd(cmdGitPath)
-	if err != nil {
-		return nil, err
-	}
-
-	emptyTreeSHA, err := git.Cmd(cmdEmptyTreeSHA)
-	if err != nil {
-		log.Debug("Couldn't get empty tree SHA value, not critical")
-	}
-
 	git.root = rootPath
 
-	return &Repository{
+	r := &Repository{
 		Fs:                fs,
 		Git:               git,
 		HooksPath:         hooksPath,
 		RootPath:          rootPath,
 		GitPath:           gitPath,
 		InfoPath:          infoPath,
 		unstagedPatchPath: filepath.Join(infoPath, unstagedPatchName),
-		emptyTreeSHA:      emptyTreeSHA,
-	}, nil
+	}
+
+	r.InitializeForTest()
+
+	return r, nil
+}
+
+// Precompute runs various Git commands in the background so the results are ready.
+// This returns a function which can be used to wait for the result. This should
+// be invoked to ensure we're not holding any locks on the Git repository.
+func (r *Repository) Precompute() func() {
+	var eg errgroup.Group
+
+	eg.Go(func() error {
+		_, _ = r.stagedFilesOnce()
+		return nil
+	})
+
+	eg.Go(func() error {
+		_, _ = r.stagedFilesWithDeletedOnce()
+		return nil
+	})
+
+	eg.Go(func() error {
+		_, _ = r.statusShortOnce()
+		return nil
+	})
+
+	return func() {
+		_ = eg.Wait()
+	}
+}
+
+// InitializeForTest must be called if you constructed a Repository without using `NewRepository`.
+// This can also be called multiple times to reset the cache.
+func (r *Repository) InitializeForTest() {
```

Comment:
Could you call this function differently, please? We use it in the New func, so it's not only for test. It can be `SetupStageCommands` or `SetupCachedCommands`, or even a mysterious `Setup` :)

Reply:
I opted for just `Setup`. In hindsight I realize that `SetupCache` is maybe more precise, but since it's all in `internal/` I'm not sure it matters.